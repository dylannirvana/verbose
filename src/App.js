/*
* Entry point for the App
 */

import React, {Component} from 'react';


import Header from './components/Header/Header';
import ProductGrid from './components/ProductGrid/ProductGrid';
import './App.scss';
import FilterFactory from "./components/GridControls/Filters/FilterFactory";
import GridControls from "./components/GridControls";
import {Container, Jumbotron} from 'reactstrap';
import Spinner from './components/Spinner';
import FileUploader from "./components/GridControls/FileUploader/FileUploader";


class App extends Component {

    constructor(props) {
        super(props);
        const component = this;

        this.state = {

            filterFactory: new FilterFactory(),

            spinnerVisibility: 'none',

            packeryRefresh: true, /// whether packery should be refreshed
            packery: false, //Reference to the Packery Instance
            dragableComponents: [] //Array of dragable product components
        };

        //Container allows the child components to manage the state of the GridOrder component
        this.container = {

            //Functions for managing state
            getState: function (state) {
                return state !== undefined ? component.state[state] : component.state
            },

            setState: function (state) {

                component.setState(state);
            },

            initFilterFactory(feed){
                component.state.filterFactory.init(feed);
                component.forceUpdate();

            },


            //Indicates if the product grid has been populated with data
            gridPopulated: function () {
                return Boolean(component.state.feed.length);
            },

            getFilterFactory: function () {

                return component.state.filterFactory;
            },

            showSpinner: () => {
                component.setState({spinnerVisibility: 'block'});
            },

            hideSpinner: () => {
                component.setState({spinnerVisibility: 'none'});
            }


        }

        return this;
    }


    render() {

        return (
            <div id={"page"} >

                <Spinner spinnerVisibility={this.state.spinnerVisibility}/>
                <Header container={this.container}/>

                <Jumbotron  id="app-intro" className={"text-center"}>
                    <Container>
                        <h1 className={"jumbotron-heading"}>Grid Order App</h1>
                        <p className={"lead text-muted"}>
                            Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
                        </p>
                    </Container>
                </Jumbotron>

                {
                    //If there are no products uploaded via CSV, display the FileUploader component
                    !this.state.filterFactory.productsAvailable() &&
                     <FileUploader container={this.container}/>
                }

                {/*Render the controls for controlling the product grid*/}
                <GridControls container={this.container}/>

                <ProductGrid container={this.container}/>
            </div>
        );
    }
}

export default App;