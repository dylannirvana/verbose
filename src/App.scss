@import 'components/Header/Header';
@import 'components/GridControls/GridControls';
@import 'components/GridControls/FileUploader/FileUploader';
@import 'components/ProductGrid/Product';

#page {


  .row, .col {
    margin: 0;
    padding: 0;
  }


}


.btn {
  border-radius: 3px;
  margin: 0;
  cursor: pointer;
  transition: 0.5s background-color, 0.5s color;
}


#app-intro{
  padding-bottom: 20px;
}

#page-content{
  background: white;
  max-width: 730px;
  width: 730px;
  margin-left: auto;
  margin-right: auto;
  margin-top: 0;
  margin-bottom: 20px;
  padding-top: 0;
}


