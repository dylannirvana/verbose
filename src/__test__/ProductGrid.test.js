import React, {Component} from 'react';
import ShallowRenderer from 'react-test-renderer/shallow'; // ES6
import Enzyme from 'enzyme';
import {createSerializer} from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import App from '../App';


//Test Data, which contains 17 products
const DATA = JSON.parse('[{"sku":"81134","name":"Studio Swing Arm Floor Lamp","category":"Floor > Task","function":"Functional","relatives":"VC CLASSIC","designer":"Studio VC","image":"https://www.circalighting.com/media/catalog/product/8/1/81134hab_5.png","gridorder":"1215.0000","neworder":"","__parsed_extra":[""]},{"sku":"82034","name":"Studio Swing Arm Wall Light","category":"Wall > Task","function":"Functional","relatives":"VC CLASSIC","designer":"Studio VC","image":"https://www.circalighting.com/media/catalog/product/8/2/82034hab_1.png","gridorder":"4742.0000","neworder":"","__parsed_extra":[""]},{"sku":"91025","name":"Studio Adjustable Floor Lamp","category":"Floor > Task","function":"Functional","relatives":"VC CLASSIC","designer":"Studio VC","image":"https://www.circalighting.com/media/catalog/product/9/1/91025hab_5.png","gridorder":"1211.0000","neworder":"","__parsed_extra":[""]},{"sku":"92000D","name":"Classic Swing Arm Wall Lamp","category":"Wall > Task","function":"Functional","relatives":"VC CLASSIC","designer":"Studio VC","image":"https://www.circalighting.com/media/catalog/product/9/2/92000dhabl_1.png","gridorder":"4536.0000","neworder":"","__parsed_extra":[""]},{"sku":"92025","name":"Studio Swing Arm Wall Light","category":"Wall > Task","function":"Functional","relatives":"VC CLASSIC","designer":"Studio VC","image":"https://www.circalighting.com/media/catalog/product/9/2/92025hab_1.png","gridorder":"4738.0000","neworder":"","__parsed_extra":[""]},{"sku":"AH2000","name":"Dean Single Arm Sconce","category":"Wall > Decorative","function":"Decorative","relatives":"Dean","designer":"Alexa Hampton","image":"https://www.circalighting.com/media/catalog/product/a/h/ah2000pnnp_5.png","gridorder":"4159.0000","neworder":"","__parsed_extra":[""]},{"sku":"AH2001","name":"Dean Library Sconce","category":"Wall > Decorative","function":"Decorative","relatives":"Dean","designer":"Alexa Hampton","image":"https://www.circalighting.com/media/catalog/product/a/h/ah2001pnnp_8.png","gridorder":"4161.0000","neworder":"","__parsed_extra":[""]},{"sku":"AH2002","name":"Abbot Single Arm Sconce","category":"Wall > Decorative","family": "test-family", "function":"Decorative","relatives":"Abbot","designer":"Alexa Hampton","image":"https://www.circalighting.com/media/catalog/product/a/h/ah2002pnnp_8.png","gridorder":"4144.0000","neworder":"","__parsed_extra":[""]},{"sku":"AH2003","name":"Abbot Library Sconce","category":"Wall > Decorative","function":"Decorative","relatives":"Abbot","designer":"Alexa Hampton","image":"https://www.circalighting.com/media/catalog/product/a/h/ah2003gmnp_5.png","gridorder":"4146.0000","neworder":"","__parsed_extra":[""]},{"sku":"AH2004","name":"Bing Single Arm Sconce","category":"Wall > Decorative","function":"Decorative","relatives":"Bing","designer":"Alexa Hampton","image":"https://www.circalighting.com/media/catalog/product/a/h/ah2004pnnp_5.png","gridorder":"4137.0000","neworder":"","__parsed_extra":[""]},{"sku":"AH2010","name":"Ginger Single Arm Sconce","category":"Wall > Decorative","function":"Decorative","relatives":"Ginger","designer":"Alexa Hampton","image":"https://www.circalighting.com/media/catalog/product/a/h/ah2010pnnp_5.png","gridorder":"4139.0000","neworder":"","__parsed_extra":[""]},{"sku":"AH2012","name":"Gene Swing Arm","category":"Wall > Task","function":"Functional","relatives":"Gene","designer":"Alexa Hampton","image":"https://www.circalighting.com/media/catalog/product/a/h/ah2012nbs_1.png","gridorder":"4525.0000","neworder":"","__parsed_extra":[""]},{"sku":"AH2013","name":"Gene Library Sconce","category":"Wall > Decorative", "family": "test-family", "function":"Decorative","relatives":"Gene","designer":"Alexa Hampton","image":"https://www.circalighting.com/media/catalog/product/a/h/ah2013nbnp_5.png","gridorder":"4150.0000","neworder":"","__parsed_extra":[""]},{"sku":"AH2700","name":"Dean 9\\" Picture Light","category":"Wall > Picture","function":"Picture","relatives":"Dean2","designer":"Alexa Hampton","image":"https://www.circalighting.com/media/catalog/product/a/h/ah2700nb_1.png","gridorder":"5324.0000","neworder":"","__parsed_extra":[""]},{"sku":"AH2701","name":"Dean 12\\" Picture Light","category":"Wall > Picture","function":"Picture","relatives":"Dean2","designer":"Alexa Hampton","image":"https://www.circalighting.com/media/catalog/product/a/h/ah2701nb_1.png","gridorder":"5328.0000","neworder":"","__parsed_extra":[""]},{"sku":"AH2702","name":"Dean 18\\" Picture Light","category":"Wall > Picture","function":"Picture","relatives":"Dean2","designer":"Alexa Hampton","image":"https://www.circalighting.com/media/catalog/product/a/h/ah2702nb_1.png","gridorder":"5332.0000","neworder":"","__parsed_extra":[""]},{"sku":"AH2703","name":"Dean 24\\" Picture Light","category":"Wall > Picture","function":"Picture","relatives":"Dean2","designer":"Alexa Hampton","image":"https://www.circalighting.com/media/catalog/product/a/h/ah2703nb_1.png","gridorder":"5336.0000","neworder":"","__parsed_extra":[""]}]');

//Configure Enzyme to work with React
Enzyme.configure({ adapter: new Adapter() })

//Add snapshot serializer to JEST
expect.addSnapshotSerializer(createSerializer({mode: 'deep'}));

//Create Renderer
const renderer = new ShallowRenderer();
renderer.render( <App  />)

//Get the App instance
const APP_INSTANCE = renderer._instance;




const FILTER_FACTORY = APP_INSTANCE.container.getFilterFactory(); //Filter Factory

FILTER_FACTORY.init(DATA);



const EXPECTED_PRODUCT_COUNT = {category: 15, function: 7, family: 7}; //Expected product count for each filter





/*
* Toggles filter option
 */
function toggleFilterOption(filterName, selectedOption, deselect = false){

    const FILTERED_PRODUCTS = FILTER_FACTORY.toggleFilterOption(filterName,selectedOption);

    if(!deselect){
        //EXPECT: Category attribute of each product to contain the keyword - Wall
        FILTERED_PRODUCTS.forEach((product) => {
            if( product[filterName] !== undefined){
                expect(product[filterName].toLowerCase()).toContain(selectedOption);
            }

        })
    }
    return FILTERED_PRODUCTS;
}




describe('Initial check',()=>{


    it(`Filter Factory populated with ${DATA.length} products`, () => {
        expect(FILTER_FACTORY.getFeed()).toHaveLength(DATA.length);
    });


})





describe('Verifying Filtering Logic',() => {
    for(let i =0; i < 3; i++){

        const SELECT = i%2 == 0 ? true : false;


        it(`${SELECT ? 'SELECTING' : 'DESELECTING'} Category -> Wall; Expected ${i%2 == 0 ? EXPECTED_PRODUCT_COUNT.category : DATA.length} products`  ,()=>{

            const FILTERED_PRODUCTS = toggleFilterOption('category','wall', !SELECT);

            //EXPECT: Correct product count is returned returned
            expect(FILTERED_PRODUCTS).toHaveLength(SELECT ? EXPECTED_PRODUCT_COUNT.category : DATA.length);



        })
    }

    for(let i =0; i < 3; i++){

        const SELECT = i%2 == 0 ? true : false;



        it(`${SELECT ? 'SELECTING' : 'DESELECTING'} Function -> Decorative; Expected ${i%2 == 0 ? EXPECTED_PRODUCT_COUNT.function : EXPECTED_PRODUCT_COUNT.category} products`  ,()=>{


            const FILTERED_PRODUCTS = toggleFilterOption('function','decorative', !SELECT);

            //EXPECT: Correct product count is returned returned
            expect(FILTERED_PRODUCTS).toHaveLength(SELECT ? EXPECTED_PRODUCT_COUNT.function : EXPECTED_PRODUCT_COUNT.category);


        })


    }
})





describe('Verifying Sorting Logic -> Sorting 2 products from the data set, with family == "test-family"',()=>{


    //Test Data has two products with family == "test-family"
    it('Verifying the two "test-family" products are not the first two products of the grid'  ,()=>{

        //Select the sort option
        const FILTERED_PRODUCTS = toggleFilterOption('family','test-family');

        //FILTER_FACTORY should be ready to sort
        const SHOULD_SORT = FILTER_FACTORY.shouldSort('family')
        expect(SHOULD_SORT).toEqual(true);


        //Before sort, first two products should not be having Family attribute's value "test-family"
        !expect(FILTERED_PRODUCTS[0].family).toEqual('test-family');
        !expect(FILTERED_PRODUCTS[1].family).toEqual('test-family');



    })

    it('After sorting, verifying the products with family == "test-family" are the first two products of the grid',()=>{
        //Do the sorting
        const SORTED_PRODUCTS = FILTER_FACTORY.sortProducts({filterName: "family", selectedOption:"test-family"});

        //Product count should be the same before and after the sorting
        expect(SORTED_PRODUCTS).toHaveLength(7);

        //After sort, first two products should  be having Family attribute's value "test-family"
        expect(SORTED_PRODUCTS[0].family).toEqual('test-family');
        expect(SORTED_PRODUCTS[1].family).toEqual('test-family');


    })


})





//Save the new order grid to a snapshot
describe('Saving New Grid Order to snapshot',() => {
    it('Snapshot created',()=>{
        expect(FILTER_FACTORY.getFilteredProducts()).toMatchSnapshot();
    })
})