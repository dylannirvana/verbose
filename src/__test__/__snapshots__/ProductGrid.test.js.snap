// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Saving New Grid Order to snapshot Snapshot created 1`] = `
Array [
  Object {
    "__parsed_extra": Array [
      "",
    ],
    "category": "Wall > Decorative",
    "designer": "Alexa Hampton",
    "family": "test-family",
    "function": "Decorative",
    "gridorder": "4144.0000",
    "image": "https://www.circalighting.com/media/catalog/product/a/h/ah2002pnnp_8.png",
    "name": "Abbot Single Arm Sconce",
    "neworder": "",
    "relatives": "Abbot",
    "sku": "AH2002",
  },
  Object {
    "__parsed_extra": Array [
      "",
    ],
    "category": "Wall > Decorative",
    "designer": "Alexa Hampton",
    "family": "test-family",
    "function": "Decorative",
    "gridorder": "4150.0000",
    "image": "https://www.circalighting.com/media/catalog/product/a/h/ah2013nbnp_5.png",
    "name": "Gene Library Sconce",
    "neworder": "",
    "relatives": "Gene",
    "sku": "AH2013",
  },
  Object {
    "__parsed_extra": Array [
      "",
    ],
    "category": "Wall > Decorative",
    "designer": "Alexa Hampton",
    "function": "Decorative",
    "gridorder": "4159.0000",
    "image": "https://www.circalighting.com/media/catalog/product/a/h/ah2000pnnp_5.png",
    "name": "Dean Single Arm Sconce",
    "neworder": "",
    "relatives": "Dean",
    "sku": "AH2000",
  },
  Object {
    "__parsed_extra": Array [
      "",
    ],
    "category": "Wall > Decorative",
    "designer": "Alexa Hampton",
    "function": "Decorative",
    "gridorder": "4161.0000",
    "image": "https://www.circalighting.com/media/catalog/product/a/h/ah2001pnnp_8.png",
    "name": "Dean Library Sconce",
    "neworder": "",
    "relatives": "Dean",
    "sku": "AH2001",
  },
  Object {
    "__parsed_extra": Array [
      "",
    ],
    "category": "Wall > Decorative",
    "designer": "Alexa Hampton",
    "function": "Decorative",
    "gridorder": "4146.0000",
    "image": "https://www.circalighting.com/media/catalog/product/a/h/ah2003gmnp_5.png",
    "name": "Abbot Library Sconce",
    "neworder": "",
    "relatives": "Abbot",
    "sku": "AH2003",
  },
  Object {
    "__parsed_extra": Array [
      "",
    ],
    "category": "Wall > Decorative",
    "designer": "Alexa Hampton",
    "function": "Decorative",
    "gridorder": "4137.0000",
    "image": "https://www.circalighting.com/media/catalog/product/a/h/ah2004pnnp_5.png",
    "name": "Bing Single Arm Sconce",
    "neworder": "",
    "relatives": "Bing",
    "sku": "AH2004",
  },
  Object {
    "__parsed_extra": Array [
      "",
    ],
    "category": "Wall > Decorative",
    "designer": "Alexa Hampton",
    "function": "Decorative",
    "gridorder": "4139.0000",
    "image": "https://www.circalighting.com/media/catalog/product/a/h/ah2010pnnp_5.png",
    "name": "Ginger Single Arm Sconce",
    "neworder": "",
    "relatives": "Ginger",
    "sku": "AH2010",
  },
]
`;
